Code/GraphMol/MarvinParse/MarvinParser.cpp

```diff
@@ -0,0 +1,1110 @@
+        MolOps::detectBondStereochemistry(*mol);
+      }
+
+      if (mol->hasProp(common_properties::_NeedsQueryScan)) {
+        mol->clearProp(common_properties::_NeedsQueryScan);
+        QueryOps::completeMolQueries(mol);
+      }
+
+      // clean up
+
+      for (auto &stereoGroup : stereoGroups) {
+        delete stereoGroup;
+      }
+
+      return mol;
+    }
+
+    catch (const std::exception &e) {
+      delete mol;
+
+      delete conf;
+
+      for (auto &stereoGroup : stereoGroups) {
+        delete stereoGroup;
+      }
+
+      throw;
+    }
+  }
+
+  MarvinMolBase *parseMarvinMolecule(
+      boost::property_tree::ptree molTree,
+      MarvinMol *parentMol = nullptr)  // parent is for sub-mols
+  {
+    MarvinMolBase *res = nullptr;
+
+    try {
+      std::string role = "";
+
+      if (parentMol == nullptr) {
+        res = new MarvinMol(molTree);
+      } else  // is is a sub-mol - used for sGroups
+      {
+        role = molTree.get<std::string>("<xmlattr>.role", "");
+        if (role == "") {
+          throw FileParseException(
+              "Expected a role for a sub-molecule in MRV file");
+        }
+
+        if (role == "SuperatomSgroup") {
+          // there are two types of superatomSgroups - regular and expanded.
+          // Expanded has a molecule attr of atomRefs
+
+          std::string atomRefs =
+              molTree.get<std::string>("<xmlattr>.atomRefs", "");
+          auto atomArray = molTree.get_child_optional("atomArray");
+
+          if (atomRefs == "" &&
+              atomArray)  // no atomRefs means regular superatom
+                          // - the atoms are in this superatom
+          {
+            res = new MarvinSuperatomSgroup(parentMol, molTree);
+
+          } else  // if atomRefs and atomArray does not exist, this is an
+                  // expanded superatom - the atoms are already in the parent
+                  // mol
+          {
+            res = new MarvinSuperatomSgroupExpanded(parentMol, molTree);
+          }
+        } else if (role == "SruSgroup" || role == "CopolymerSgroup" ||
+                   role == "ModificationSgroup") {
+          res = new MarvinSruCoModSgroup(parentMol, role, molTree);
+        } else if (role == "MultipleSgroup") {
+          res = new MarvinMultipleSgroup(parentMol, molTree);
+        } else if (role == "MulticenterSgroup") {
+          res = new MarvinMulticenterSgroup(parentMol, molTree);
+        } else if (role == "GenericSgroup") {
+          res = new MarvinGenericSgroup(parentMol, molTree);
+        } else if (role == "MonomerSgroup") {
+          res = new MarvinMonomerSgroup(parentMol, molTree);
+        } else if (role == "DataSgroup") {
+          res = new MarvinDataSgroup(parentMol, molTree);
+        } else {
+          throw FileParseException("Unexpected role " + role + " in MRV file");
+        }
+      }
+
+      BOOST_FOREACH (boost::property_tree::ptree::value_type &v, molTree) {
```

Comment:
Why use BOOST_FOREACH here (and below) instead of a standard range for loop ?

Reply:
```suggestion
      for (auto &v : molTree) {
```

Reply:
the reason:  I found an example using the BOOST_FOREACH, and it worked.  I never looked back.

Reply:
I changed them all now

Reply:
I changed them all